File: projeto1/src/broker/Coordinator.py
import zmq
import msgpack
from datetime import datetime
import threading
import time

# ============ RELÓGIO LÓGICO ============
logical_clock = 0

def clock_before_send() -> int:
    global logical_clock
    logical_clock += 1
    print(f"[COORDENADOR][RELÓGIO] Incrementado antes de enviar: {logical_clock}", flush=True)
    return logical_clock

def clock_on_receive(received: int) -> None:
    global logical_clock
    old = logical_clock
    if received > logical_clock:
        logical_clock = received + 1
        print(f"[COORDENADOR][RELÓGIO] ATUALIZADO! Recebido={received} > local={old}. Novo={logical_clock}", flush=True)
    else:
        logical_clock += 1
        print(f"[COORDENADOR][RELÓGIO] MANTIDO! Recebido={received} <= local={old}. Novo={logical_clock}", flush=True)


# ============ ESTADO DOS SERVIDORES ============
# { nome: { "rank": int, "last_heartbeat": float } }
servers: dict = {}
servers_lock = threading.Lock()
next_rank = 1
HEARTBEAT_TIMEOUT = 30  # segundos


# ============ HANDLERS ============

def process_get_rank(msg: dict) -> dict:
    global next_rank
    server_name = msg.get("username", "").strip()
    if not server_name:
        return {"success": False, "message": "Nome do servidor inválido", "logical_clock": 0, "rank": 0, "current_time": 0}

    with servers_lock:
        if server_name in servers:
            rank = servers[server_name]["rank"]
            print(f"[COORDENADOR] Servidor já cadastrado: {server_name} (rank={rank})", flush=True)
        else:
            rank = next_rank
            next_rank += 1
            servers[server_name] = {
                "rank": rank,
                "last_heartbeat": time.time(),
            }
            print(f"[COORDENADOR] NOVO servidor cadastrado: {server_name} (rank={rank})", flush=True)

    return {"success": True, "message": "Rank atribuído", "rank": rank, "logical_clock": 0, "current_time": 0}


def process_get_server_list(msg: dict) -> dict:
    with servers_lock:
        server_list = [
            {"name": name, "rank": info["rank"]}
            for name, info in servers.items()
        ]
    print(f"[COORDENADOR] Retornando lista com {len(server_list)} servidores", flush=True)
    for s in server_list:
        print(f"  - {s['name']} (rank={s['rank']})", flush=True)
    return {"success": True, "message": "Lista de servidores", "server_list": server_list, "logical_clock": 0, "current_time": 0, "rank": 0}


def process_heartbeat(msg: dict) -> dict:
    server_name = msg.get("username", "").strip()
    with servers_lock:
        if server_name not in servers:
            return {"success": False, "message": "Servidor não cadastrado", "logical_clock": 0, "rank": 0, "current_time": 0}
        servers[server_name]["last_heartbeat"] = time.time()

    current_time = int(datetime.now().timestamp() * 1000)
    print(f"[COORDENADOR] Heartbeat de: {server_name} (rank={servers[server_name]['rank']})", flush=True)
    return {"success": True, "message": "Heartbeat OK", "current_time": current_time, "logical_clock": 0, "rank": 0}


# ============ THREAD DE LIMPEZA ============

def cleanup_loop():
    while True:
        time.sleep(10)
        now = time.time()
        with servers_lock:
            inativos = [
                name for name, info in servers.items()
                if now - info["last_heartbeat"] > HEARTBEAT_TIMEOUT
            ]
            for name in inativos:
                rank = servers.pop(name)["rank"]
                print(f"[COORDENADOR] Servidor REMOVIDO por inatividade: {name} (rank={rank})", flush=True)


# ============ MAIN ============
context = zmq.Context()
socket = context.socket(zmq.REP)
socket.bind("tcp://*:5560")

print("COORDENADOR PYTHON INICIADO", flush=True)
print("Porta: 5560", flush=True)
print("Aguardando servidores...\n", flush=True)

t = threading.Thread(target=cleanup_loop, daemon=True)
t.start()

while True:
    try:
        raw = socket.recv()
        msg = msgpack.unpackb(raw, raw=False)

        clock_on_receive(msg.get("logical_clock", 0))
        print(f"\n[COORDENADOR] Requisição: {msg.get('type')} de {msg.get('username')}", flush=True)

        msg_type = msg.get("type")
        if msg_type == "get_rank":
            response = process_get_rank(msg)
        elif msg_type == "get_server_list":
            response = process_get_server_list(msg)
        elif msg_type == "heartbeat":
            response = process_heartbeat(msg)
        else:
            response = {"success": False, "message": "Tipo desconhecido", "logical_clock": 0, "rank": 0, "current_time": 0}

        response["logical_clock"] = clock_before_send()
        socket.send(msgpack.packb(response))

    except Exception as e:
        print(f"[COORDENADOR] Erro: {e}", flush=True)
        err = {"success": False, "message": "Erro no coordenador", "logical_clock": clock_before_send(), "rank": 0, "current_time": 0}
        socket.send(msgpack.packb(err))

File: projeto1/src/broker/Coordinator.java
import org.zeromq.ZMQ;
import org.zeromq.ZContext;
import java.util.*;
import java.util.concurrent.*;

// gerencia servidores (rank, lista, heartbeat, sincronização)
class Coordinator {
    private static long logicalClock = 0;
    
    // PARTE 3: Mapa de servidores: nome -> ServerRecord
    private static Map<String, ServerRecord> servers = new ConcurrentHashMap<>();
    
    // PARTE 3: Contador para atribuir ranks únicos
    private static int nextRank = 1;
    
    // PARTE 3: Timeout para considerar servidor inativo (30 segundos)
    private static final long HEARTBEAT_TIMEOUT = 30000;
    
    public static void main(String[] args) {
        try (ZContext context = new ZContext()) {
            // Socket REP para receber requisições dos servidores
            ZMQ.Socket socket = context.createSocket(ZMQ.REP);
            socket.bind("tcp://*:5559");
            
            System.out.println("COORDENADOR INICIADO");
            System.out.println("Porta: 5559");
            System.out.println("Aguardando servidores...\n");
            
            // Thread para limpar servidores inativos
            startCleanupThread();
            
            while (true) {
                try {
                    // Receber mensagem do servidor
                    byte[] msgBytes = socket.recv(0);
                    Message msg = MessagePackUtil.deserialize(msgBytes, Message.class);
                    
                    // PARTE 3: AO RECEBER - atualizar relógio lógico
                    long receivedClock = msg.getLogicalClock();
                    long oldClock = logicalClock;
                    
                    if (receivedClock > logicalClock) {
                        logicalClock = receivedClock + 1;
                        System.out.println("[COORDENADOR][RELÓGIO] ATUALIZADO! Recebido=" + receivedClock + 
                                         " > local=" + oldClock + ". Novo=" + logicalClock);
                    } else {
                        logicalClock++;
                        System.out.println("[COORDENADOR][RELÓGIO] MANTIDO! Recebido=" + receivedClock + 
                                         " <= local=" + oldClock + ". Novo=" + logicalClock);
                    }
                    
                    System.out.println("\n[COORDENADOR] Requisição: " + msg.getType() + " de " + msg.getUsername());
                    
                    Response response;
                    
                    // Processar baseado no tipo
                    if ("get_rank".equals(msg.getType())) {
                        // PARTE 3: Atribuir rank (enunciado linha 14)
                        response = processGetRank(msg);
                    } else if ("get_server_list".equals(msg.getType())) {
                        // PARTE 3: Retornar lista de servidores (enunciado linha 22)
                        response = processGetServerList(msg);
                    } else if ("heartbeat".equals(msg.getType())) {
                        // PARTE 3: Processar heartbeat (enunciado linha 27 e 33)
                        response = processHeartbeat(msg);
                    } else {
                        response = new Response(false, "Tipo desconhecido");
                    }
                    
                    // PARTE 3: ANTES DE ENVIAR - incrementar relógio
                    logicalClock++;
                    response.setLogicalClock(logicalClock);
                    System.out.println("[COORDENADOR][RELÓGIO] Incrementado antes de enviar: " + logicalClock);
                    
                    // Enviar resposta
                    byte[] responseBytes = MessagePackUtil.serialize(response);
                    socket.send(responseBytes, 0);
                    
                } catch (Exception e) {
                    System.err.println("[COORDENADOR] Erro: " + e.getMessage());
                    Response errorResponse = new Response(false, "Erro no coordenador");
                    logicalClock++;
                    errorResponse.setLogicalClock(logicalClock);
                    byte[] errorBytes = MessagePackUtil.serialize(errorResponse);
                    socket.send(errorBytes, 0);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    
    // PARTE 3: Processar requisição de rank (enunciado linha 14)
    private static Response processGetRank(Message msg) {
        String serverName = msg.getUsername();
        
        if (serverName == null || serverName.trim().isEmpty()) {
            return new Response(false, "Nome do servidor inválido");
        }
        
        ServerRecord server;
        
        // PARTE 3: Verificar se servidor já tem rank (enunciado linha 19 - sem repetições)
        if (servers.containsKey(serverName)) {
            server = servers.get(serverName);
            System.out.println("[COORDENADOR] Servidor já cadastrado: " + serverName + " (rank=" + server.getRank() + ")");
        } else {
            // PARTE 3: Atribuir novo rank
            server = new ServerRecord(serverName, nextRank++);
            servers.put(serverName, server);
            System.out.println("[COORDENADOR] NOVO servidor cadastrado: " + serverName + " (rank=" + server.getRank() + ")");
        }
        
        Response response = new Response(true, "Rank atribuído");
        response.setRank(server.getRank());
        return response;
    }
    
    // PARTE 3: Processar requisição de lista (enunciado linha 22)
    private static Response processGetServerList(Message msg) {
        List<ServerRecord> serverList = new ArrayList<>(servers.values());
        
        System.out.println("[COORDENADOR] Retornando lista com " + serverList.size() + " servidores");
        for (ServerRecord s : serverList) {
            System.out.println("  - " + s.getName() + " (rank=" + s.getRank() + ")");
        }
        
        Response response = new Response(true, "Lista de servidores");
        response.setServerList(serverList);
        return response;
    }
    
    // PARTE 4: Processar heartbeat (sem retornar hora - eleição cuida da sincronização)
    private static Response processHeartbeat(Message msg) {
        String serverName = msg.getUsername();
        
        if (serverName == null || !servers.containsKey(serverName)) {
            return new Response(false, "Servidor não cadastrado");
        }
        
        // Atualizar timestamp do último heartbeat
        ServerRecord server = servers.get(serverName);
        server.setLastHeartbeat(System.currentTimeMillis());
        
        System.out.println("[COORDENADOR] Heartbeat de: " + serverName + " (rank=" + server.getRank() + ")");
        
        // PARTE 4: Não retorna mais currentTime - sincronização é feita via eleição
        Response response = new Response(true, "Heartbeat OK");
        return response;
    }
    
    // Thread para remover servidores inativos 
    private static void startCleanupThread() {
        Thread cleanupThread = new Thread(() -> {
            while (true) {
                try {
                    Thread.sleep(10000); 
                    
                    long currentTime = System.currentTimeMillis();
                    List<String> inactiveServers = new ArrayList<>();
                    
                    // Identificar servidores inativos
                    for (Map.Entry<String, ServerRecord> entry : servers.entrySet()) {
                        ServerRecord server = entry.getValue();
                        long timeSinceLastHeartbeat = currentTime - server.getLastHeartbeat();
                        
                        if (timeSinceLastHeartbeat > HEARTBEAT_TIMEOUT) {
                            inactiveServers.add(entry.getKey());
                        }
                    }
                    
                    // Remover servidores inativos
                    for (String serverName : inactiveServers) {
                        ServerRecord removed = servers.remove(serverName);
                        System.out.println("[COORDENADOR] Servidor REMOVIDO por inatividade: " + 
                                         serverName + " (rank=" + removed.getRank() + ")");
                    }
                    
                } catch (InterruptedException e) {
                    break;
                }
            }
        });
        
        cleanupThread.setDaemon(true);
        cleanupThread.start();
    }
}
